### PR TITLE
Add location information to `WarnIfLowProductivity`

### LINKED ISSUE #2528: WarnIfLowProductivity should provide some location information
```
     [exec] kore-exec: [30559389] Warning (WarnIfLowProductivity):
     [exec]     Productivity dropped to: 40%
     [exec]     Poor productivity may indicate a performance bug.
     [exec]     Please file a bug report: https://github.com/kframework/kore/issues
```
Currently, I'm getting a few of these on Jenkins and I don't know which tests.

This error message should be accompanied by at least something that could pinpoint the original file used to run kore-exec.

### PR DESCRIPTION
**Pull request by andreiburdusa**
Fixes #2528 

---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [ ] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [ ] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [ ] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

### CODE REVIEW

File: kore/app/exec/Main.hs [hunk 1]
```diff
@@ -587,10 +590,13 @@ mainWithOptions execOptions = do
                         }
             writeOptionsAndKoreFiles tmpDir execOptions'
             e <-
-                mainDispatch execOptions' <* warnIfLowProductivity
-                    & handle handleWithConfiguration
-                    & handle handleSomeException
-                    & runKoreLog tmpDir koreLogOptions
+                let run = do
+                        (kFileLocations, exitCode) <- mainDispatch execOptions'
+                        warnIfLowProductivity kFileLocations
```
Comment (ttuegel) [inline kore/app/exec/Main.hs:595]: I think we should move `warnIfLowProductivity` down into `mainDispatch`.

File: kore/app/share/GlobalMain.hs [hunk 2]
```diff
@@ -478,13 +484,19 @@ type LoadedModule = VerifiedModule Attribute.Symbol
 
 type LoadedDefinition = (Map ModuleName LoadedModule, Map Text AstLocation)
```
Comment (ttuegel) [inline kore/app/share/GlobalMain.hs:485]: I think we should make this a record data type and embed the `KFileLocations` there.

File: kore/src/Kore/Attribute/SourceLocation.hs [hunk 3]
```diff
@@ -52,6 +53,8 @@ data SourceLocation = SourceLocation
 instance Default SourceLocation where
     def = SourceLocation def def
 
+notDefault :: SourceLocation -> Bool
+notDefault = (/=) def
```
Comment (ttuegel) [inline kore/src/Kore/Attribute/SourceLocation.hs:57]: Where is this used?

File: kore/src/Kore/Log/WarnIfLowProductivity.hs [hunk 4]
```diff
@@ -16,30 +17,50 @@ import Pretty (
...
-
+                    <> kFiles
+                    <> [ "Poor productivity may indicate a performance bug."
+                       , "Please file a bug report: https://github.com/kframework/kore/issues"
+                       ]
+          where
+            kFiles
+                | not . null $ locations =
+                    [ Pretty.nest 4 $
+                        Pretty.vsep
```
Comment (ttuegel) [inline kore/src/Kore/Log/WarnIfLowProductivity.hs:46]: Better phrasing here will fix the indentation:

```suggestion
                    [ (Pretty.nest 4 . Pretty.vsep)
```

File: kore/src/Kore/Log/WarnIfLowProductivity.hs [hunk 5]
```diff
@@ -16,30 +17,50 @@ import Pretty (
...
+                    ]
                 ]
-            , "Poor productivity may indicate a performance bug."
-            , "Please file a bug report: https://github.com/kframework/kore/issues"
-            ]
-
+                    <> kFiles
+                    <> [ "Poor productivity may indicate a performance bug."
+                       , "Please file a bug report: https://github.com/kframework/kore/issues"
+                       ]
```
Comment (ttuegel) [inline kore/src/Kore/Log/WarnIfLowProductivity.hs:41]: Let's use `concat` here.

**Review by ttuegel — CHANGES_REQUESTED** [review]

File: kore/src/Kore/Attribute/SourceLocation.hs (hunk 3 above)
Comment (andreiburdusa) [reply kore/src/Kore/Attribute/SourceLocation.hs:57]: Here https://github.com/kframework/kore/pull/2594/files#diff-10bb26c5f6fb3d7794de9cbf4fc75e448ce311e8673315c27ffb7b0ef7f98a9eR497

File: kore/app/exec/Main.hs [hunk 6]
```diff
@@ -624,20 +629,27 @@ mainDispatch :: KoreExecOptions -> Main ExitCode
...
+            else koreProve execOptions proveOptions & warnProductivity
     | Just searchOptions <- koreSearchOptions =
-        koreSearch execOptions searchOptions
+        koreSearch execOptions searchOptions & warnProductivity
     | Just mergeOptions <- koreMergeOptions =
-        koreMerge execOptions mergeOptions
+        koreMerge execOptions mergeOptions & warnProductivity
     | otherwise =
-        koreRun execOptions
+        koreRun execOptions & warnProductivity
```
Comment (ana-pantilie) [inline kore/app/exec/Main.hs:639]: Would it be better if we had a `mainDispatchWorker` and `mainDispatch` would look like `warnProductivity mainDispatchWorker` or something similar? The idea is that we call `warnProductivity` on every branch now, and we could reduce duplication.

File: kore/src/Kore/Repl/Data.hs [hunk 7]
```diff
@@ -577,6 +578,7 @@ data ReplState = ReplState
     , -- | The log level, log scopes and log type decide what gets logged and
       -- where.
       koreLogOptions :: !Log.KoreLogOptions
+    , kFileLocations :: KFileLocations
```
Comment (ana-pantilie) [inline kore/src/Kore/Repl/Data.hs:581]: Is there any reason this was added to the `State` and not the `Config`?

**Review by ttuegel — APPROVED** [review]

**Pull request merged**